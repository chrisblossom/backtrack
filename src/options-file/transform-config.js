/* @flow */

import ExConfig from 'ex-config';
import { Preprocessor } from './preprocessor';
import { postProcessor } from './post-processor';
import { presetValidator } from './preset-validator';
import { cleanPreprocessor } from './clean-preprocessor';
import { cleanValidator } from './clean-validator';
import { cleanProcessor } from './clean-processor';
import { filesPreprocessor } from './files-preprocessor';
import { filesValidator } from './files-validator';

import { rootPath } from '../config/paths';

import type { Lifecycles, Preset } from '../types.js';

function transformConfig(
    config: Preset,
    dirname?: string = rootPath,
): Lifecycles {
    /**
     * Needs to be placed inside function for caching reasons
     */
    const options = {
        preprocessor: new Preprocessor(),
        postProcessor,
        plugins: false,
        validator: presetValidator,
        processor: 'arrayPush',
        overrides: {
            presets: {
                resolve: {
                    prefix: 'backtrack-preset',
                    org: '@backtrack',
                    orgPrefix: 'preset',
                },
            },
            clean: {
                preprocessor: cleanPreprocessor,
                validator: cleanValidator,
                processor: cleanProcessor,
            },
            files: {
                preprocessor: filesPreprocessor,
                validator: filesValidator,
            },
            resolve: {
                processor: 'mergeDeep',
            },
        },
    };

    const exConfig = new ExConfig(options);

    const result: Lifecycles = exConfig.load(config, dirname);

    return result;
}

export { transformConfig };
