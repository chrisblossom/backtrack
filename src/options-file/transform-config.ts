import { exConfigSync, Options as exConfigOptions } from 'ex-config';
import { Preprocessor } from './preprocessor';
import { postProcessor } from './post-processor';
import { presetValidator } from './preset-validator';
import { cleanPreprocessor } from './clean-preprocessor';
import { cleanValidator } from './clean-validator';
import { cleanProcessor } from './clean-processor';
import { filesPreprocessor } from './files-preprocessor';
import { filesValidator } from './files-validator';

import { rootPath } from '../config/paths';

import { Lifecycles, Preset } from '../types';

function transformConfig(
    config: Preset,
    dirname: string = rootPath,
): Lifecycles {
    /**
     * Needs to be placed inside function for caching reasons
     */
    const options: exConfigOptions = {
        baseDirectory: dirname,
        preprocessor: Preprocessor(),
        postProcessor,
        plugins: false,
        validator: presetValidator,
        processor: 'arrayPush',
        overrides: {
            presets: {
                resolve: {
                    prefix: 'backtrack-preset',
                    org: '@backtrack',
                    orgPrefix: 'preset',
                },
            },
            clean: {
                preprocessor: cleanPreprocessor,
                validator: cleanValidator,
                processor: cleanProcessor,
            },
            files: {
                preprocessor: filesPreprocessor,
                validator: filesValidator,
            },
            resolve: {
                processor: 'mergeDeep',
            },
        },
    };

    const result: Lifecycles = exConfigSync(config, options);

    return result;
}

export { transformConfig };
