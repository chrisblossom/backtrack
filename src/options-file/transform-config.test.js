/* @flow */

import path from 'path';

const transformConfig = (config, dirname) =>
    require('./transform-config').transformConfig(config, dirname);

describe('transformConfig', () => {
    const cwd = process.cwd();

    afterEach(() => {
        process.chdir(cwd);
    });

    test('handles one preset', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const config = {
            presets: ['01'],
            clean: [
                {
                    del: ['**/*', '!.gitignore'],
                    makeDirs: ['static/favicons'],
                    copy: [
                        {
                            src: 'static-3',
                            dest: 'static-3',
                        },
                    ],
                },
            ],
            dev: ['base'],
        };

        const result = transformConfig(config);

        expect(result).toMatchSnapshot();
    });

    test('handles false', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const config = {
            presets: ['01'],
            clean: false,
        };

        const result = transformConfig(config);

        expect(result).toMatchSnapshot();
    });

    test('handles false with additional task', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const config = {
            presets: ['01'],
            clean: [false, { del: 'static' }],
        };

        const result = transformConfig(config);

        expect(result).toMatchSnapshot();
    });

    test('handles false in middle of preset chain', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const config = {
            presets: ['03'],
            dev: ['base'],
        };

        const result = transformConfig(config);

        expect(result).toMatchSnapshot();
    });

    test('handles false in middle of preset chain with additional task', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const config = {
            presets: ['04'],
            dev: ['base'],
        };

        const result = transformConfig(config);

        expect(result).toMatchSnapshot();
    });

    test('handles false in middle of long preset chain', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const config = {
            presets: ['05'],
            dev: ['base'],
        };

        const result = transformConfig(config);

        expect(result).toMatchSnapshot();
    });
});
