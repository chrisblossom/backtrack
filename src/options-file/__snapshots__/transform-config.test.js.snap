// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformConfig handles false 1`] = `
Object {
  "dev": Array [
    [Function preset2DevTask1],
    Object {
      "name": "preset1DevTask1",
      "task": [Function preset1DevTask1],
    },
    Object {
      "name": "run jest",
      "task": "jest",
    },
    "eslint --fix",
    Array [
      Object {
        "name": "run prettier",
        "task": "prettier",
      },
      "eslint",
    ],
    [Function preset1DevTask2],
    Array [
      [Function preset1DevTask1],
      [Function preset1DevTask2],
    ],
  ],
  "files": Object {
    "dest": Object {
      "absolute": Object {},
      "allowChanges": Object {},
      "files": Array [],
      "hash": Object {},
    },
    "makeDirs": Array [
      "dist",
      "src",
    ],
    "src": Object {
      "absolute": Object {},
      "files": Array [],
      "hash": Object {},
    },
  },
  "resolve": Object {
    "backtrack-preset-01": "<PROJECT_ROOT>/node_modules/backtrack-preset-01",
    "backtrack-preset-02": "<PROJECT_ROOT>/node_modules/backtrack-preset-02",
  },
}
`;

exports[`transformConfig handles false in middle of long preset chain 1`] = `
Object {
  "dev": Array [
    "preset-06",
    "preset-05",
    "base",
  ],
  "files": Object {
    "dest": Object {
      "absolute": Object {},
      "allowChanges": Object {},
      "files": Array [],
      "hash": Object {},
    },
    "makeDirs": Array [
      "dist",
      "src",
    ],
    "src": Object {
      "absolute": Object {},
      "files": Array [],
      "hash": Object {},
    },
  },
  "resolve": Object {
    "backtrack-preset-02": "<PROJECT_ROOT>/node_modules/backtrack-preset-02",
    "backtrack-preset-05": "<PROJECT_ROOT>/node_modules/backtrack-preset-05",
    "backtrack-preset-06": "<PROJECT_ROOT>/node_modules/backtrack-preset-06",
    "backtrack-preset-07": "<PROJECT_ROOT>/node_modules/backtrack-preset-07",
  },
}
`;

exports[`transformConfig handles false in middle of preset chain 1`] = `
Object {
  "dev": Array [
    "base",
  ],
  "files": Object {
    "dest": Object {
      "absolute": Object {},
      "allowChanges": Object {},
      "files": Array [],
      "hash": Object {},
    },
    "makeDirs": Array [
      "dist",
      "src",
    ],
    "src": Object {
      "absolute": Object {},
      "files": Array [],
      "hash": Object {},
    },
  },
  "resolve": Object {
    "backtrack-preset-01": "<PROJECT_ROOT>/node_modules/backtrack-preset-03",
    "backtrack-preset-02": "<PROJECT_ROOT>/node_modules/backtrack-preset-02",
  },
}
`;

exports[`transformConfig handles false in middle of preset chain with additional task 1`] = `
Object {
  "dev": Array [
    "preset-04",
    "base",
  ],
  "files": Object {
    "dest": Object {
      "absolute": Object {},
      "allowChanges": Object {},
      "files": Array [],
      "hash": Object {},
    },
    "makeDirs": Array [
      "dist",
      "src",
    ],
    "src": Object {
      "absolute": Object {},
      "files": Array [],
      "hash": Object {},
    },
  },
  "resolve": Object {
    "backtrack-preset-02": "<PROJECT_ROOT>/node_modules/backtrack-preset-02",
    "backtrack-preset-04": "<PROJECT_ROOT>/node_modules/backtrack-preset-04",
  },
}
`;

exports[`transformConfig handles false with additional task 1`] = `
Object {
  "clean": Object {
    "copy": Array [],
    "del": Array [
      "static",
    ],
    "makeDirs": Array [],
  },
  "dev": Array [
    [Function preset2DevTask1],
    Object {
      "name": "preset1DevTask1",
      "task": [Function preset1DevTask1],
    },
    Object {
      "name": "run jest",
      "task": "jest",
    },
    "eslint --fix",
    Array [
      Object {
        "name": "run prettier",
        "task": "prettier",
      },
      "eslint",
    ],
    [Function preset1DevTask2],
    Array [
      [Function preset1DevTask1],
      [Function preset1DevTask2],
    ],
  ],
  "files": Object {
    "dest": Object {
      "absolute": Object {},
      "allowChanges": Object {},
      "files": Array [],
      "hash": Object {},
    },
    "makeDirs": Array [
      "dist",
      "src",
    ],
    "src": Object {
      "absolute": Object {},
      "files": Array [],
      "hash": Object {},
    },
  },
  "resolve": Object {
    "backtrack-preset-01": "<PROJECT_ROOT>/node_modules/backtrack-preset-01",
    "backtrack-preset-02": "<PROJECT_ROOT>/node_modules/backtrack-preset-02",
  },
}
`;

exports[`transformConfig handles one preset 1`] = `
Object {
  "clean": Object {
    "copy": Array [
      Object {
        "dest": "<PROJECT_ROOT>/dist/static-1",
        "src": "<PROJECT_ROOT>/static-1",
      },
      Object {
        "dest": "<PROJECT_ROOT>/dist/static-2",
        "hash": true,
        "src": "<PROJECT_ROOT>/static-2",
      },
      Object {
        "dest": "<PROJECT_ROOT>/dist/static-3",
        "src": "<PROJECT_ROOT>/static-3",
      },
    ],
    "del": Array [
      "nested",
      "**/*",
      "!.gitignore",
    ],
    "makeDirs": Array [
      "<PROJECT_ROOT>/dist/static/other",
      "<PROJECT_ROOT>/dist/static/favicons",
    ],
  },
  "dev": Array [
    [Function preset2DevTask1],
    Object {
      "name": "preset1DevTask1",
      "task": [Function preset1DevTask1],
    },
    Object {
      "name": "run jest",
      "task": "jest",
    },
    "eslint --fix",
    Array [
      Object {
        "name": "run prettier",
        "task": "prettier",
      },
      "eslint",
    ],
    [Function preset1DevTask2],
    Array [
      [Function preset1DevTask1],
      [Function preset1DevTask2],
    ],
    "base",
  ],
  "files": Object {
    "dest": Object {
      "absolute": Object {},
      "allowChanges": Object {},
      "files": Array [],
      "hash": Object {},
    },
    "makeDirs": Array [
      "dist",
      "src",
    ],
    "src": Object {
      "absolute": Object {},
      "files": Array [],
      "hash": Object {},
    },
  },
  "resolve": Object {
    "backtrack-preset-01": "<PROJECT_ROOT>/node_modules/backtrack-preset-01",
    "backtrack-preset-02": "<PROJECT_ROOT>/node_modules/backtrack-preset-02",
  },
}
`;
